src/java.base/share/classes/com/sun/crypto/provider/PBES2Core.java

```diff
@@ -184,89 +202,59 @@ protected void engineInit(int opmode, Key key,
                 throw new InvalidKeyException("Missing password");
             }
 
-            // TBD: consolidate the salt, ic and IV parameter checks below
-
-            // Extract salt and iteration count from the key, if present
-            if (key instanceof javax.crypto.interfaces.PBEKey) {
-                salt = ((javax.crypto.interfaces.PBEKey)key).getSalt();
-                if (salt != null && salt.length < 8) {
+            boolean doEncrypt = ((opmode == Cipher.ENCRYPT_MODE) ||
+                        (opmode == Cipher.WRAP_MODE));
+
+            // Extract from the supplied PBE params, if present
+            if (params instanceof PBEParameterSpec pbeParams) {
+                // salt should be non-null per PBEParameterSpec
+                salt = check(pbeParams.getSalt());
+                iCount = check(pbeParams.getIterationCount());
+                AlgorithmParameterSpec ivParams = pbeParams.getParameterSpec();
+                if (ivParams instanceof IvParameterSpec iv) {
+                    ivSpec = iv;
+                } else if (ivParams == null && doEncrypt) {
+                    // generate random IV
+                    byte[] ivBytes = new byte[blkSize];
+                    random.nextBytes(ivBytes);
+                    ivSpec = new IvParameterSpec(ivBytes);
+                } else {
                     throw new InvalidAlgorithmParameterException(
-                            "Salt must be at least 8 bytes long");
+                            "Wrong parameter type: IV expected");
                 }
-                iCount = ((javax.crypto.interfaces.PBEKey)key).getIterationCount();
-                if (iCount == 0) {
-                    iCount = DEFAULT_COUNT;
-                } else if (iCount < 0) {
-                    throw new InvalidAlgorithmParameterException(
-                            "Iteration count must be a positive number");
+            } else if (params == null && doEncrypt) {
+                // Try extracting from the key, if present
+                if (key instanceof javax.crypto.interfaces.PBEKey pbeKey) {
+                    salt = check(pbeKey.getSalt()); // may be null
+                    iCount = check(pbeKey.getIterationCount()); // may be 0
```

Comment:
It seems the return value is never 0.

Reply:
Oh, the comment about iteration may be 0 is meant toward the pbeKey.getInterationCount() call... Hmm, I will make it clearer.

Reply:
I see. Another question, shall we reset `salt` and `iCount` at the beginning? If `params` is null and `key` is not `PBEKey` and there is an existing positive `iCount`, it will not be set to `DEFAULT_COUNT`.

Reply:
Yeah, I see PKCS12PBECipherCore does that. Will update PBES2Core with it too.